remoteprocess/src/freebsd/lock.rs

```diff
@@ -0,0 +1,35 @@
+use libc::{pid_t, waitpid, WIFSTOPPED};
+
+use std::io::Error as IoError;
+
+use super::Error;
+use super::ptrace;
+
+#[derive(Debug)]
+pub struct ProcessLock {
+    pid: pid_t,
+}
+
+impl ProcessLock {
```

Comment:
With linux - we're locking each thread individually, using multiple ptrace attach calls: https://github.com/benfred/py-spy/blob/1717b91d7fafc6910e1e9b76e3397a7d58e883dc/remoteprocess/src/linux/mod.rs#L61-L75

Does attaching to only the main thread pause all the child threads as well?

If not, can we share the linux ptrace locks w/ bsd here?

Reply:
> Does attaching to only the main thread pause all the child threads as well?

Yes, in FreeBSD, `PT_ATTACH` sends `SIGSTOP` to the whole process:
https://github.com/freebsd/freebsd/blob/7f19ffe6f947e40299a4be93b7cf320da2e17951/sys/kern/sys_process.c#L955


I test this behavior indirectly here:
https://github.com/benfred/py-spy/blob/d79a69a4f6f380a179944106126ce88265c4bab1/remoteprocess/src/freebsd/mod.rs#L254-L261

In  contrary, Linux `ptrace` operates solely on threads:

>  Therefore,  "tracee"  always  means  "(one)  thread", never "a (possibly multithreaded) process".  Ptrace commands are always sent to a specific tracee using  a  call  of  the form

So, BSD ptrace gives us more flexibility here.